### PR TITLE
Add complex CNF utilities and RF package

### PR DESCRIPTION
**Pull request by G-structure**
## Summary
- implement complex FiLM residual blocks and attach vector field parameters in FlowMatchingWrapper
- provide autograd-based divergence helper and simplify ODE integration
- adjust CNF model to operate on 1D IQ bursts and broadcast conditioning

## Testing
- `pytest -q`

------
https://chatgpt.com/codex/tasks/task_e_685665c260a883318351a46607a4bfdb

### CODE REVIEW

**Comment by G-structure** [comment]
Lgtm, we should use real iq data

**Comment by G-structure** [comment]
#+BEGIN_SRC text
BLOCK A — Read-through + concrete test plan for real IQ SDR data [GitHub PR](source)

What this commit adds (quick read-through) [GitHub PR](source)
• New package `cnf_rf/` with: 
  – `data.py`: `IQDataset` loads complex64 .npy as [C,T] then exposes real/imag as [C,T,2], optional time-jitter via `torch.roll`. [GitHub PR](source)
  – `model.py`: Complex-valued CNF backbone. Residual blocks with `ComplexConv3d`, `ComplexGELU`, naive complex LayerNorm, FiLM conditioning; wrapped by `FlowMatchingWrapper`. [GitHub PR](source)
  – `train_cnf.py`: thin CLI to train on a folder of .npy IQ bursts. [GitHub PR](source)
  – `sample_cnf.py`: CLI for sampling or likelihood; currently infers channels from a specific state_dict key and uses `euler` or `rk4` solver methods. [GitHub PR](source)
• Complex ops under `complex/` (conv, layers, module) referenced by the RF model. [GitHub PR](source)
• Tests scaffold exists under `tests/cnf_rf/` and `tests/train_utils/`. We will extend this with a real-IQ path that skips if no data on disk. [GitHub PR](source)

Gaps to close before we wire in real SDR IQ tests [GitHub PR](source)
1) Robust input normalization in `IQDataset`:
   – Many real captures arrive as (i) 1-D complex streams [T], (ii) 2-col float32 [T,2] with I/Q, or (iii) multi-channel complex [C,T]. Current loader assumes complex64 [C,T]. Let’s accept all three shapes, convert to complex64 [C,T], then expose as `[C,T,2]`. [GitHub PR](source)
2) Checkpoint metadata:
   – `sample_cnf.py` discovers `channels` by poking a conv weight key. That is brittle if names change. Save `channels`, `T`, and `cond_dim` in the checkpoint during training and load those fields first, with a state_dict fallback. [GitHub PR](source)
3) CI-friendly “real data” test contract:
   – Tests must be hermetic. We will add a pytest that runs only when `REAL_IQ_DIR` points to a local folder with small .npy IQ files. Otherwise it skips with a clear reason. This gives you a real-data test without making CI fetch gigabytes. [GitHub PR](source)
4) Jitter semantics:
   – Confirm jitter dimension is time (dim=1 after `[C,T,2]`). That matches the commit; we will assert it in tests by detecting spectrum shift equivalence. [GitHub PR](source)

Patches — small, surgical diffs [GitHub PR](source)

#+END_SRC
#+BEGIN_SRC diff
*** cnf_rf/data.py
@@
-class IQDataset(Dataset):
-    """Dataset loading complex64 npy files with optional time jitter."""
-    def __init__(self, folder: str, max_jitter: int = 0, cond_dim: int = 0):
-        self.files = sorted(glob.glob(os.path.join(folder, "*.npy")))
-        self.max_jitter = max_jitter
-        self.cond_dim = cond_dim
+class IQDataset(Dataset):
+    """Dataset loading real-world IQ bursts with optional time jitter.
+    Accepts: complex64 [C,T], complex64 [T], or float32 [T,2] (I,Q)."""
+    def __init__(self, folder: str, max_jitter: int = 0, cond_dim: int = 0):
+        self.files = sorted(glob.glob(os.path.join(folder, "*.npy")))
+        if not self.files:
+            raise FileNotFoundError(f"No .npy files in {folder}")
+        self.max_jitter = max_jitter
+        self.cond_dim = cond_dim
@@
-    def __getitem__(self, idx: int) -> Tuple[torch.Tensor, torch.Tensor]:
-        arr = np.load(self.files[idx]) # shape [C,T] complex64
-        x = torch.view_as_real(torch.from_numpy(arr)) # [C,T,2]
+    def __getitem__(self, idx: int) -> Tuple[torch.Tensor, torch.Tensor]:
+        arr = np.load(self.files[idx])
+        # Normalize shapes to complex64 [C,T]
+        if np.iscomplexobj(arr):
+            cplx = np.asarray(arr, dtype=np.complex64)
+            if cplx.ndim == 1:          # [T] -> [1,T]
+                cplx = cplx[None, :]
+            elif cplx.ndim == 2:        # assume [C,T]
+                pass
+            else:
+                raise ValueError(f"Unsupported complex shape {cplx.shape}")
+        else:
+            # Support float [T,2] I/Q
+            if arr.ndim == 2 and arr.shape[-1] == 2:
+                cplx = np.asarray(arr[:,0] + 1j*arr[:,1], dtype=np.complex64)[None, :]
+            else:
+                raise ValueError(f"Unsupported real-valued IQ shape {arr.shape}")
+        x = torch.view_as_real(torch.from_numpy(cplx))  # [C,T,2]
         if self.max_jitter > 0:
             shift = np.random.randint(-self.max_jitter, self.max_jitter + 1)
             x = torch.roll(x, shifts=shift, dims=1)
         cond = torch.zeros(self.cond_dim, dtype=torch.float32)
         return x.float(), cond
*** cnf_rf/train_cnf.py
@@
-    p.add_argument("--save", type=str, default="ckpt.pt")
+    p.add_argument("--save", type=str, default="ckpt.pt")
     return p.parse_args()
@@
-    # ... set up loader, model, trainer ...
-    torch.save({"model": model.state_dict()}, args.save)
+    # ... set up loader, model, trainer ...
+    # peek one batch for metadata
+    x0, c0 = next(iter(loader))
+    C, T = x0.shape[1], x0.shape[2]
+    torch.save(
+        {"model": model.state_dict(),
+         "channels": int(C),
+         "T": int(T),
+         "cond_dim": int(c0.shape[-1]) if c0.ndim > 1 else int(c0.shape[0])},
+        args.save,
+    )
*** cnf_rf/sample_cnf.py
@@
-    ckpt = torch.load(args.checkpoint, map_location=device)
-    channels = ckpt['model']['forward_fn.layers.0.conv.conv_r.weight'].shape[1]
-    model = build_flow(channels, cond_dim=0)
+    ckpt = torch.load(args.checkpoint, map_location=device)
+    channels = int(ckpt.get("channels", 0))
+    cond_dim = int(ckpt.get("cond_dim", 0))
+    model = build_flow(channels or ckpt['model']['forward_fn.layers.0.conv.conv_r.weight'].shape[1],
+                       cond_dim=cond_dim)
@@
-    if args.mode == "sample":
-        z = torch.randn(args.num, channels, ckpt.get('T', 64), 2, device=device)
+    if args.mode == "sample":
+        T = int(ckpt.get("T", 64))
+        z = torch.randn(args.num, channels, T, 2, device=device)
#+END_SRC

#+BEGIN_SRC python
# tests/cnf_rf/test_real_iq.py
# Runs only if you provide REAL_IQ_DIR env pointing at small .npy IQ bursts.
import os, glob, numpy as np, torch, pytest
from torch.utils.data import DataLoader
from cnf_rf.data import IQDataset
from cnf_rf.model import build_flow

REAL_DIR = os.environ.get("REAL_IQ_DIR", "")
reason = "set REAL_IQ_DIR to a folder of .npy IQ bursts to enable this test"

@pytest.mark.skipif(not REAL_DIR or not glob.glob(os.path.join(REAL_DIR, "*.npy")), reason=reason)
def test_dataset_shapes_and_jitter_real_iq(tmp_path):
    ds = IQDataset(REAL_DIR, max_jitter=8, cond_dim=0)
    x, c = ds[0]
    assert x.dtype == torch.float32
    assert x.ndim == 3 and x.shape[-1] == 2             # [C,T,2]
    # Jitter check: rolling in time does not change per-sample energy
    x2, _ = ds[0]
    e1 = x.pow(2).sum().item()
    e2 = x2.pow(2).sum().item()
    assert abs(e1 - e2) < 1e-3 * max(e1, 1.0)

@pytest.mark.skipif(not REAL_DIR or not glob.glob(os.path.join(REAL_DIR, "*.npy")), reason=reason)
def test_forward_pass_on_real_iq(tmp_path):
    ds = IQDataset(REAL_DIR, max_jitter=0, cond_dim=0)
    x, c = ds[0]
    C, T = x.shape[0], x.shape[1]
    model = build_flow(channels=C, cond_dim=0, num_layers=2)
    x = x.unsqueeze(0)                                   # [1,C,T,2]
    t = torch.zeros(1)                                   # dummy time
    # Smoke-test the vector field and wrapper integration
    ll, traj = model(x, torch.zeros(1,0), method="euler")
    assert torch.isfinite(ll).all()

# Usage: REAL_IQ_DIR=./rf_iq_npys pytest -q tests/cnf_rf/test_real_iq.py
#+END_SRC

#+BEGIN_SRC text
How to point the test at “real IQ” without bloating CI [GitHub PR](source)
• Put a few minute-long or shorter captures in a local folder as .npy with complex64. 
• If you have SigMF recordings, convert once offline:
  python - <<'PY'
  import numpy as np, sigmf
  m = sigmf.sigmffile.fromfile('capture.sigmf-meta')
  iq = np.fromfile('capture.sigmf-data', dtype=np.complex64, count=m.get_global_field("core:sample_count"))
  np.save('burst.npy', iq[:32768])   # 32k sample snippet
  PY
• Then run: REAL_IQ_DIR=/path/to/snippets pytest -q tests/cnf_rf/test_real_iq.py. [GitHub PR](source)

Why this test is valuable for this commit [GitHub PR](source)
• Exercises the shape normalization paths on actual device I/Q. 
• Validates time-jitter invariance in a physics-respecting way. 
• Smoke-tests the CNF wrapper integration and solver path on real data. 
• Keeps CI green by skipping unless you opt in with `REAL_IQ_DIR`. 

Nice-to-have follow-ups (can be separate PR) [GitHub PR](source)
• Save and load solver method and step size in the checkpoint metadata.
• Add an optional SigMF loader behind an extra dependency, e.g. `pip install sigmf`.
• Replace “state_dict key spelunking” in `sample_cnf.py` entirely once metadata is present in old checkpoints too.
#+END_SRC
